napari/utils/notifications.py

```diff
@@ -0,0 +1,220 @@
+            self.INFO: "ⓘ",
+            self.NONE: "",
+        }[self]
+
+
+ActionSequence = Sequence[Tuple[str, Callable[[], None]]]
+
+
+class Notification(Event):
+    """A Notifcation event.  Usually created by :class:`NotificationManager`.
+
+    Parameters
+    ----------
+    message : str
+        The main message/payload of the notification.
+    severity : str or NotificationSeverity, optional
+        The severity of the notification, by default
+        `NotificationSeverity.WARNING`.
+    actions : sequence of tuple, optional
+        Where each tuple is a `(str, callable)` 2-tuple where the first item
+        is a name for the action (which may, for example, be put on a button),
+        and the callable is a callback to perform when the action is triggered.
+        (for example, one might show a traceback dialog). by default ()
+    type : str, optional
+        The notification event type, by default 'notification'.  May be changed
+        by subclasses.
+    native : Any, optional
+        A native backend event that may have triggered this Notification,
+        by default None
+    """
+
+    def __init__(
+        self,
+        message: str,
+        severity: Union[
+            str, NotificationSeverity
+        ] = NotificationSeverity.WARNING,
+        # source: Optional[str] = None, # TODO
+        actions: ActionSequence = (),
+        type: str = 'notification',
+        native: Any = None,
+        **kwargs,
+    ):
+
+        super().__init__(type, **kwargs)
+        self.message = message
+        self.severity = NotificationSeverity(severity)
+        # self.source = source  # TODO
+        self.actions = actions
+
+        # let's store when the object was created;
+        self.date = datetime.now()
+
+    @classmethod
+    def from_exception(cls, exc: BaseException, **kwargs) -> 'Notification':
+        return ErrorNotification(exc, **kwargs)
+
+    @classmethod
+    def from_warning(cls, warning: Warning, **kwargs) -> 'Notification':
+        return WarningNotification(warning, **kwargs)
+
+
+class ErrorNotification(Notification):
+    exception: BaseException
+
+    def __init__(self, exception: BaseException, *args, **kwargs):
+        msg = getattr(exception, 'message', str(exception))
+        severity = getattr(exception, 'severity', 'ERROR')
+        actions = getattr(exception, 'actions', ())
+        super().__init__(msg, severity, actions, type='error')
+        self.exception = exception
+
+
+class WarningNotification(Notification):
+    warning: Warning
+
+    def __init__(self, warning: Warning, *args, **kwargs):
+        msg = getattr(warning, 'message', str(warning))
+        severity = getattr(warning, 'severity', 'WARNING')
+        actions = getattr(warning, 'actions', ())
+        super().__init__(msg, severity, actions, type='warning')
+        self.warning = warning
+
+
+class NotificationManager:
+    """
+    A Singleton instance of notification manager.
```

Comment:
Sorry if I've missed the discussion, but do we want one notification manager even in the case of multiple viewers?

Reply:
I don't see why we wouldn't. The goal is to collect all notification in a single place. You also said in your main comments that we can't yet see the traceback, and indeed, I removed that code as we decided to separate the Purely Qt part into a separate PR; And IIUC the question of manager per viewer is more a question of where you want to see the notification than a question of how many notification manager we want to have.

Reply:
for now @jni, yes.  The question of whether a notification gets shown _visually_ in one of the viewers would be handled later... and could probably be handled by a filter that determines whether the _source_ of the `Notification` (since notifications are just `Event` objects) is related to a specific viewer.  but all for another PR.